libvast/src/system/count_command.cpp

```diff
@@ -0,0 +1,124 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/system/count_command.hpp"
+
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/expression.hpp"
+#include "vast/defaults.hpp"
+#include "vast/error.hpp"
+#include "vast/logger.hpp"
+#include "vast/scope_linked.hpp"
+#include "vast/system/atoms.hpp"
+#include "vast/system/signal_monitor.hpp"
+#include "vast/system/spawn_or_connect_to_node.hpp"
+#include "vast/system/start_command.hpp"
+#include "vast/system/tracker.hpp"
+
+#include <caf/actor.hpp>
+#include <caf/event_based_actor.hpp>
+#include <caf/scoped_actor.hpp>
+#include <caf/settings.hpp>
+#include <caf/stateful_actor.hpp>
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+namespace vast::system {
+
+caf::message
+count_command(const command::invocation& invocation, caf::actor_system& sys) {
+  VAST_TRACE(invocation);
+  auto first = invocation.arguments.begin();
+  auto last = invocation.arguments.end();
+  const auto& options = invocation.options;
+  // Read query from input file, STDIN or CLI arguments.
+  std::string query;
+  auto assign_query = [&](std::istream& in) {
+    query.assign(std::istreambuf_iterator<char>{in},
+                 std::istreambuf_iterator<char>{});
+  };
+  if (auto fname = caf::get_if<std::string>(&options, "count.read")) {
```

Comment:
This whole dispatching logic seems very similar to `detail::make_(input|output)_stream`. I wonder if we can reuse this logic here.

Reply:
I'm not quite sure I grok `make_input_stream` after looking at it for a bit. Do you mean something along the lines of:

```c++
if (auto is = detail::make_input_stream<detail::count>(options))
  assign_query(*is);
```

Reply:
This code does already exist in `sink_command.cpp:52-90`, `make_input_stream` is something different.

Reply:
Before we create a separate task and the whole ceremony for this, lets just consolidate the functionality within this PR. To make the request more explicit: Please factor the shared logic between `count_command` and `sink_command` into a dedicated function that will be used by both.

Reply:
I implemented this as requested and factored it into a shared utility function named `vast::system::read_query` (for the lack of a better name).